cpp/src/io/functions.cpp

```diff
@@ -214,29 +215,80 @@ std::vector<std::vector<std::string>> read_orc_statistics(source_info const& src
   orc::metadata metadata(source.get());
 
   // Initialize statistics to return
-  std::vector<std::vector<std::string>> statistics_blobs;
+  raw_orc_statistics result;
 
   // Get column names
-  statistics_blobs.emplace_back();
   for (auto i = 0; i < metadata.get_num_columns(); i++) {
-    statistics_blobs.back().push_back(metadata.get_column_name(i));
+    result.column_names.push_back(metadata.get_column_name(i));
   }
 
   // Get file-level statistics, statistics of each column of file
-  statistics_blobs.emplace_back();
   for (auto const& stats : metadata.ff.statistics) {
-    statistics_blobs.back().push_back(std::string(stats.cbegin(), stats.cend()));
+    result.column_stats.push_back(std::string(stats.cbegin(), stats.cend()));
   }
 
   // Get stripe-level statistics
   for (auto const& stripe_stats : metadata.md.stripeStats) {
-    statistics_blobs.emplace_back();
+    result.stripe_stats.emplace_back();
     for (auto const& stats : stripe_stats.colStats) {
-      statistics_blobs.back().push_back(std::string(stats.cbegin(), stats.cend()));
+      result.stripe_stats.back().push_back(std::string(stats.cbegin(), stats.cend()));
     }
   }
 
-  return statistics_blobs;
+  return result;
+}
+
+void set_column_statistics_type(column_statistics* cs)
+{
+  if (cs->intStatistics.get()) {
+    cs->type = statistics_type::INT;
+  } else if (cs->doubleStatistics.get()) {
+    cs->type = statistics_type::DOUBLE;
+  } else if (cs->stringStatistics.get()) {
+    cs->type = statistics_type::STRING;
+  } else if (cs->bucketStatistics.get()) {
+    cs->type = statistics_type::BUCKET;
+  } else if (cs->decimalStatistics.get()) {
+    cs->type = statistics_type::DECIMAL;
+  } else if (cs->dateStatistics.get()) {
+    cs->type = statistics_type::DATE;
+  } else if (cs->binaryStatistics.get()) {
+    cs->type = statistics_type::BINARY;
+  } else if (cs->timestampStatistics.get()) {
+    cs->type = statistics_type::TIMESTAMP;
+  }
```

Comment:
The fact that everything is prefixed with `cs->` makes me think this should be a member function of `column_statistics`

Reply:
In fact, just do this from the constructor of `column_statistics`. The constructor would have to take something returned from the `ProtobufReader::read`, but that would allow eliminating the ITM pattern below.

Reply:
If I'm understanding this correctly, I should use separate types for reading protobuf and the actual return value from the `read_parsed_orc_statistics` call:
The "public" type would be constructible from the internal one. 
The type that we return to the user should deal with type erasure better than `column_statistics` currently does.